component/rbac.jsonnet

```diff
@@ -54,7 +55,231 @@ local aggregated_rbac = [
+// argocd-operator's `CONTROLLER_CLUSTER_ROLE` and `SERVER_CLUSTER_ROLE`, cf.
+// https://argocd-operator.readthedocs.io/en/latest/usage/custom_roles/
+// NOTE(sg): we only deploy these cluster roles if the respective
+// `cluster_role_match_labels` parameter in `operator` isn't empty.
+local internalControllerAggregationLabel = {
+  'rbac.argocd.syn.tools/aggregate-to-controller': 'true',
+};
+
+local custom_controller_cr =
+  kube.ClusterRole(common.custom_cluster_role_name('application-controller')) {
+    metadata+: {
+      annotations+: {
+        'syn.tools/description': |||
+          Custom ClusterRole which is used to give the ArgoCD application
+          controller access to managed namespaces. Intended to be configured
+          as `CONTROLLER_CLUSTER_ROLE` on the operator.
+
+          This ClusterRole aggregates the rules of all clusterroles which
+          match one of the selectors in `aggregationRule.clusterRoleSelectors`.
+        |||,
+      },
+    },
+    rules:: [],
+    aggregationRule: {
+      clusterRoleSelectors: [
+        { matchLabels: internalControllerAggregationLabel },
+      ] + [
+        sel
+        for sel in
+          std.objectValues(params.operator.controller_cluster_role_selectors)
+        if sel != null
+      ],
+    },
+  };
+
+local controller_required_cr =
+  kube.ClusterRole('syn:argocd-application-controller:required') {
+    metadata+: {
+      labels+: internalControllerAggregationLabel,
+      annotations+: {
+        'syn.tools/description': |||
+          Custom ClusterRole which is used to aggregate a basic set of
+          permissions to the custom ArgoCD application controller clusterrole
+          `%s`.
+
+          This rule is required to allow the application controller to run
+          normally (emitting events, updating application status, etc)
+        ||| % common.custom_cluster_role_name('application-controller'),
+      },
+    },
+    rules: [
+      // NOTE(sg): I wasn't able to fully determine why this is required, but
+      // the application controller logs errors when it doesn't have
+      // permissions to read secrets in target namespaces.
+      {
+        apiGroups: [ '' ],
+        resources: [ 'secrets' ],
+        verbs: [
+          'get',
+          'list',
+          'watch',
+        ],
+      },
+      // Required to access and update argocd apps which the controller is
+      // managing.
+      {
+        apiGroups: [ 'argoproj.io' ],
+        resources: [
+          'applications',
+          'applicationsets',
+          'appprojects',
+        ],
+        verbs: [
+          'create',
+          'get',
+          'list',
+          'watch',
+          'update',
+          'delete',
+          'patch',
+        ],
+      },
+      // Required to emit events for app state changes and similar
+      {
+        apiGroups: [ '' ],
+        resources: [ 'events' ],
+        verbs: [ 'create', 'list' ],
```

Comment:
Usually™️ controller-runtime also needs `patch`. Might be worth to also add it.

Reply:
added in https://github.com/projectsyn/component-argocd/pull/231/commits/c685df098aab66e2b178514c9912a6dbfddde903